src/jvmMain/kotlin/com/charleskorn/kaml/YamlNodeReader.kt

```diff
@@ -194,6 +198,37 @@ internal actual class YamlNodeReader(
         return resolvedNode.withPath(path.withAliasReference(anchor.value, event.location).withAliasDefinition(anchor.value, resolvedNode.location))
     }
 
+    private fun ensureNoDuplicateKeys(node: YamlNode) {
+        accumulateDuplicateKeys(listOf(node))
+            .takeIf { it.isNotEmpty() }
+            ?.mapValues { (_, duplicates) -> duplicates.sortedWith(compareBy({ it.endLocation.line }, { it.endLocation.column }))  }
+            ?.toSortedMap(compareBy { it.endLocation.line })
+            ?.let { throw DuplicateKeysException(it) }
+    }
+
+    private tailrec fun accumulateDuplicateKeys(
+        nodes: List<YamlNode>,
+        duplicates: MutableMap<YamlPath, List<YamlPath>> = mutableMapOf()
+    ): Map<YamlPath, List<YamlPath>> {
+        val nestedNodes = nodes.flatMap { node ->
+            when (node) {
+                is YamlList -> node.items
+                is YamlMap -> {
+                    duplicates.putAll(node.duplicates)
+                    node.entries.values
+                }
+                is YamlTaggedNode -> listOf(node.innerNode)
+                is YamlScalar, is YamlNull -> emptyList()
+            }
+        }
+
+        return if (nestedNodes.isNotEmpty()) {
+            accumulateDuplicateKeys(nestedNodes, duplicates)
+        } else {
+            duplicates
+        }
+    }
```

Comment:
This iteration could be avoided but I didn't want `YamlMap` to accept an additional parameter for collecting duplicates.